pydis_site/apps/forms/authentication.py

```diff
@@ -0,0 +1,167 @@
+        return self.admin
+
+    def refresh_data(self) -> None:
+        """Fetches user data from discord, and updates the instance."""
+        self.member = discord.get_member(self.payload["id"])
+
+        if self.member:
+            self.payload = self.member.user.dict()
+        else:
+            self.payload = discord.fetch_user_details(self.decoded_token.get("token"))
+
+        updated_info = self.decoded_token
+        updated_info["user_details"] = self.payload
+
+        self.token = encode_jwt(updated_info)
+
+
+class AuthenticationResult(typing.NamedTuple):
+    """Return scopes that the user has authenticated with."""
+
+    scopes: tuple[str, ...]
+
+
+# See https://www.django-rest-framework.org/api-guide/authentication/#custom-authentication
+class JWTAuthentication(BaseAuthentication):
+    """Custom DRF authentication backend for JWT."""
+
+    @staticmethod
+    def get_token_from_cookie(cookie: str) -> str:
+        """Parse JWT token from cookie."""
+        try:
+            prefix, token = cookie.split()
+        except ValueError:
+            msg = "Unable to split prefix and token from authorization cookie."
+            raise AuthenticationFailed(msg)
+
+        if prefix.upper() != "JWT":
+            msg = f"Invalid authorization cookie prefix '{prefix}'."
+            raise AuthenticationFailed(msg)
+
+        return token
+
+    def authenticate(
+        self,
+        request: HttpRequest,
+    ) -> tuple[FormsUser, None] | None:
+        """Handles JWT authentication process."""
+        cookie = request.COOKIES.get("token")
+        if not cookie:
+            return None
+
+        token = self.get_token_from_cookie(cookie)
+
+        try:
+            # New key.
+            payload = jwt.decode(token, settings.SECRET_KEY, algorithms=["HS256"])
+        except jwt.InvalidTokenError:
+            try:
+                # Old key. Should be removed at a certain point.
+                payload = jwt.decode(token, settings.FORMS_SECRET_KEY, algorithms=["HS256"])
+            except jwt.InvalidTokenError as e:
+                raise AuthenticationFailed(str(e))
+
+        scopes = ["authenticated"]
+
+        if not payload.get("token"):
+            msg = "Token is missing from JWT."
+            raise AuthenticationFailed(msg)
+        if not payload.get("refresh"):
+            msg = "Refresh token is missing from JWT."
+            raise AuthenticationFailed(msg)
+
+        try:
+            user_details = payload.get("user_details")
+            if not user_details or not user_details.get("id"):
+                msg = "Improper user details."
+                raise AuthenticationFailed(msg)
+        except Exception:
+            msg = "Could not parse user details."
+            raise AuthenticationFailed(msg)
+
+        user = FormsUser(
+            token,
+            user_details,
+            discord.get_member(user_details["id"]),
```

Comment:
Do we need error handling for this call or is it fine to let the exception bubble up?